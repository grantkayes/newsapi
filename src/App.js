import React, { useState, useEffect } from "react"
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faNewspaper } from '@fortawesome/free-solid-svg-icons'
import Articles from "./components/articles"
import Pagination from "./components/pagination"
import './App.css'

const App = () => { 

  //replace with your own API key or env file
  const url = `https://newsapi.org/v2/top-headlines?country=us&category=sports&apiKey=${process.env.REACT_APP_NEWS_API_KEY}`

  //React Hooks to set pagination and loaded articles
  const [articles, setArticles] = useState([])
  const [currentPage, setCurrentPage] = useState(1)
  const [articlesPerPage] = useState(3)
  const paginate = (pageNumber) => setCurrentPage(pageNumber)

  //the Hook equivalent of componentDidMount
  useEffect(() => {
    fetch(url)
    .then(res => res.json())
    .then((data) => {
      setArticles(data.articles)
    })
  }, [])

  //get the indices for pagination
  const indexOfLastArticle = currentPage * articlesPerPage
  const indexOfFirstArticle = indexOfLastArticle - articlesPerPage
  const currentArticles = articles.slice(indexOfFirstArticle, indexOfLastArticle)

  return (
    <div>
      <center className="d-block mt-3"><h1 className="font-weight-bold text-white">Newsie<FontAwesomeIcon icon={faNewspaper} size="lg" className="ml-3"/></h1></center>
      <center><p className="text-white">Your Source for Sports Newsâ„¢</p></center>
      <Articles articles={currentArticles}/>
      <Pagination articlesPerPage={articlesPerPage} totalArticles={articles.length} paginate={paginate}/>  
    </div>
  )
}

export default App
