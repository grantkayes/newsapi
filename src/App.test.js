import React from 'react';
import renderer from 'react-test-renderer'
import { mount, shallow } from 'enzyme'
import App, {articles, paginate, totalArticles, articlesPerPage, setCurrentPage} from './App'
import Articles from './components/articles'
import Pagination from './components/pagination'

describe('App', () => {
  test('app renders', () => {
    const component = renderer.create(<App />);
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  });

  it('renders one instance of articles', () => {
    const wrapper = mount(<App />)
    expect(wrapper.find(Articles).length).toEqual(1)
  });

  it('renders one instance of pagination', () => {
    const wrapper = mount(<App />)
    expect(wrapper.find(Pagination).length).toEqual(1)
  });
});

describe('Articles', () => {
  test('articles render correctly', () => {
    const articles = [{
      "source": {
      "id": null,
      "name": "Merkur.de"
      },
      "author": null,
      "title": "Rewe-Kundin macht Foto von Bio-Produkten - Kurioses Detail sehen viele spät | Verbraucher - fehmarn24.de",
      "description": "Eine Kundin machte in einer Rewe-Filiale eine ungewöhnliche Entdeckung. Als sie ein Foto davon auf Facebook postete, sorgte dieses für Verwirrung unter den Nutzern.",
      "url": "https://www.merkur.de/verbraucher/rewe-fund-foto-detail-kundin-ekel-heuschrecke-facebook-kommentare-erkennen-zr-13395011.html",
      "urlToImage": "https://www.fehmarn24.de/bilder/2019/12/31/13395011/2145860025-dieses-foto-hat-ein-user-in-einer-rewe-filile-gemacht-haetten-sie-tierische-detail-gleich-gesehen-EnoLPiTZGef.jpg",
      "publishedAt": "2020-01-05T21:58:53Z",
      "content": "Eine Kundin machte in einer Rewe-Filiale eine ungewöhnliche Entdeckung. Als sie ein Foto davon auf Facebook postete, sorgte dieses für Verwirrung unter den Nutzern.\r\n<ul><li>Eine Kundin machte in einer Rewe-Filiale einen Fund, der einigen wohl sauer aufstoßen… [+2814 chars]"
    }]

    const wrapper = mount(<Articles articles={ articles }/>)
    expect(wrapper.find('a').text()).toEqual("Rewe-Kundin macht Foto von Bio-Produkten - Kurioses Detail sehen viele spät | Verbraucher - fehmarn24.de")
  });
});

describe('Pagination', () => {
  test('pagination renders the correct number of pages', () => {
    const wrapper = mount(<div><Pagination totalArticles={20} articlesPerPage={2} paginate={(pageNumber) => setCurrentPage(pageNumber)} /></div>)
    expect(wrapper.find('.page-link').length).toEqual(10)
  });
});