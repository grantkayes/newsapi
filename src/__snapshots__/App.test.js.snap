// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App app renders 1`] = `
<div>
  <center
    className="d-block mt-3"
  >
    <h1
      className="font-weight-bold text-white"
    >
      Newsie
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-newspaper fa-w-18 fa-lg ml-3"
        data-icon="newspaper"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 576 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M552 64H88c-13.255 0-24 10.745-24 24v8H24c-13.255 0-24 10.745-24 24v272c0 30.928 25.072 56 56 56h472c26.51 0 48-21.49 48-48V88c0-13.255-10.745-24-24-24zM56 400a8 8 0 0 1-8-8V144h16v248a8 8 0 0 1-8 8zm236-16H140c-6.627 0-12-5.373-12-12v-8c0-6.627 5.373-12 12-12h152c6.627 0 12 5.373 12 12v8c0 6.627-5.373 12-12 12zm208 0H348c-6.627 0-12-5.373-12-12v-8c0-6.627 5.373-12 12-12h152c6.627 0 12 5.373 12 12v8c0 6.627-5.373 12-12 12zm-208-96H140c-6.627 0-12-5.373-12-12v-8c0-6.627 5.373-12 12-12h152c6.627 0 12 5.373 12 12v8c0 6.627-5.373 12-12 12zm208 0H348c-6.627 0-12-5.373-12-12v-8c0-6.627 5.373-12 12-12h152c6.627 0 12 5.373 12 12v8c0 6.627-5.373 12-12 12zm0-96H140c-6.627 0-12-5.373-12-12v-40c0-6.627 5.373-12 12-12h360c6.627 0 12 5.373 12 12v40c0 6.627-5.373 12-12 12z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </h1>
  </center>
  <center>
    <p
      className="text-white"
    >
      Your Source for Sports News™
    </p>
  </center>
  <table
    className="table table-dark"
  >
    <thead>
      <tr>
        <th
          scope="col"
        >
          Article Title
        </th>
        <th
          scope="col"
        >
          Source
        </th>
        <th
          scope="col"
        >
          Author
        </th>
        <th
          scope="col"
        >
          Description
        </th>
        <th
          scope="col"
        />
      </tr>
    </thead>
    <tbody />
  </table>
  <nav
    className="ml-2"
  >
    <ul
      className="pagination"
    />
  </nav>
</div>
`;

exports[`Articles articles render 1`] = `
<table
  className="table table-dark"
>
  <thead>
    <tr>
      <th
        scope="col"
      >
        Article Title
      </th>
      <th
        scope="col"
      >
        Source
      </th>
      <th
        scope="col"
      >
        Author
      </th>
      <th
        scope="col"
      >
        Description
      </th>
      <th
        scope="col"
      />
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a
          href="https://www.merkur.de/verbraucher/rewe-fund-foto-detail-kundin-ekel-heuschrecke-facebook-kommentare-erkennen-zr-13395011.html"
          target="_blank"
        >
          Rewe-Kundin macht Foto von Bio-Produkten - Kurioses Detail sehen viele spät | Verbraucher - fehmarn24.de
        </a>
      </td>
      <td>
        Merkur.de
      </td>
      <td>
        -
      </td>
      <td>
        Eine Kundin machte in einer Rewe-Filiale eine ungewöhnliche Entdeckung. Als sie ein Foto davon auf Facebook postete, sorgte dieses für Verwirrung unter den Nutzern.
      </td>
      <td>
        <img
          alt="article"
          src="https://www.fehmarn24.de/bilder/2019/12/31/13395011/2145860025-dieses-foto-hat-ein-user-in-einer-rewe-filile-gemacht-haetten-sie-tierische-detail-gleich-gesehen-EnoLPiTZGef.jpg"
        />
      </td>
    </tr>
  </tbody>
</table>
`;
